// lib/readtools.js
var Promise = require('lie');
var file_exists = require('is-there');
var _ = {
	merge: require('lodash.merge'),
};
// ----------------------------------------------------------------------------

module.exports = function readconf(cwd, basename) {
	var toolsfile = cwd + '/' + basename + '.js';

	if (file_exists(toolsfile)) {
		tools = require(toolsfile)
	}
	else { // no tools file
		tools = {};
	}

	return _.merge({}, {

		export: function (cwd, args, conf, valk) {
			// extra settings
		},

		send: function (cwd, args, conf, valk) {
			// extra settings
		},

		build: function (cwd, args, conf, type, category, host) {
			return Promise.resolve();
		},

		check: function (cwd, args, conf, type, category, host) {
			return Promise.resolve();
		},

		clean: function (cwd, args, conf) {
			return Promise.resolve();
		},

		confdryrun: function (cwd, args, conf, rsync, ssh, host) {
			// extra settings
		},

		confdeploy: function (cwd, args, conf, rsync, ssh, host) {
			// extra settings
		},

		confdiff: function (cwd, args, conf, rsync, ssh, host) {
			// extra settings
		}

	}, tools);
}
